synapse/handlers/federation.py

```diff
@@ -830,6 +830,47 @@ def maybe_backfill(self, room_id, current_depth):
             logger.debug("Not backfilling as no extremeties found.")
             return
 
+        # We only want to paginate if we can actually see the events we'll get,
+        # as otherwise we'll just spend a lot of resources to get redacted
+        # events.
+        #
+        # We do this by filtering all the extremities and seeing if any remain.
+        # Given we don't have the extremity events themselves, we need to
+        # actually check the events that reference them.
+        #
+        # *Note*: the spec wants us to keep backfilling until we reach the start
+        # of the room in case we are allowed to see some of the history. However
+        # in practice that causes more issues than its worth, as a) its
+        # relatively rare for there to be any visible history and b) even when
+        # there is its often sufficiently long ago that clients would stop
+        # attempting to paginate before backfill reached the visible history.
+        #
+        # TODO: If we do do a backfill the we should filter the extremities to
+        #   only include those that point to visible portions of history.
+        #
+        # TODO: Correctly handle the case where we are allowed to see the
+        #   forward event but not the extremity, e.g. in the case of initial
+        #   join of the server where we are allowed to see the join event but
+        #   not anything before it.
+
+        forward_events = yield self.store.get_forward_events(
+            list(extremities),
+        )
+
+        extremities_events = yield self.store.get_events(
+            forward_events,
+            check_redacted=False,
+            get_prev_content=False,
+        )
+
+        filtered_extremities = yield filter_events_for_server(
```

Comment:
I don't think filter_events_for_server is quite what we want here. There are reasons that an individual event might be filtered whilst the rest of the timeline is still visible (eg, gdpr erasure).

I fear you need to extract the bits of filter_events_for_server that are actually relevant here. I think you might also find you can skip a bunch of it.

Reply:
You're absolutely right, I've now made it only check history visibility. I haven't factored it out as it seems like a relatively minor change to add it as a parameter to the function. OTOH I can get on board with factoring stuff out if you think its getting a bit too unwieldy

Reply:
I still wonder if a function dedicated to this usecase could end up being simpler, but this is fine.